examples/bri-3/src/bri/identity/bpiSubjects/agents/bpiSubjects.agent.ts

```diff
@@ -1,22 +1,52 @@
-import { BadRequestException, Injectable } from "@nestjs/common";
+import { BadRequestException, Injectable, NotFoundException } from "@nestjs/common";
 import { BpiSubject } from "../models/bpiSubject";
 import { BpiSubjectType } from "../models/bpiSubjectType.enum";
 
 import { uuid } from 'uuidv4'; 
+import { BpiSubjectStorageAgent } from "../persistence/bpiSubjectsStorage.agent";
+import { NotFoundError } from "rxjs";
  
+// Agent methods have extremely declarative names and perform a single task 
 @Injectable()
 export class BpiSubjectAgent {
-  // Agent methods have extremely declarative names and perform a single task 
+
+  constructor(private repo: BpiSubjectStorageAgent) { }
   public throwIfCreateBpiSubjectInputInvalid(name: string, desc: string, pk: string) {
     // This is just an example, these fields will be validated on the DTO validation layer
     // This validation would check internal business rules (i.e. bpiSubject must have public key in the format defined by the participants..) 
     if (!name) {
-      // We stop execution in case of critical errors by throwing a simple exception
       throw new BadRequestException("Name cannot be empty.")
     };
   }
 
   public createNewExternalBpiSubject(name :string, description: string, publicKey: string): BpiSubject {
     return new BpiSubject(uuid(), name, description, BpiSubjectType.External, publicKey );
   }
+
+  public async fetchUpdateCandidateAndThrowIfUpdateValidationFails(id: string, name: string, desc: string, pk: string): Promise<BpiSubject> {
+    const bpiSubjectToUpdate = await this.repo.getBpiSubjectById(id);
+
+    if(!bpiSubjectToUpdate) {
+      throw new NotFoundException(`Bpi Subject with id: ${id} does not exist.`)
+    }
+
+    return bpiSubjectToUpdate;
+  }
+
+  public updateBpiSubject(bpiSubjectToUpdate: BpiSubject, name :string, description: string, publicKey: string) {
```

Comment:
the non-empty validation is done in the class definition, right?

Reply:
Yes. We will have one level of protection (i.e. from external inputs) through input DTO validation (#520) and another level of protection (i.e. from accidental bugs introduced by devs) in the domain object. Finally the DB will reject the update if the constrainst are not followed.

For point 3 we ll introduce global exception handling in the upcoming days.